### ISSUE TITLE
Compartment with new intrinsics

### ISSUE THREAD

**Issue by AidenRourke** [Agoric/SES-shim#593]
Hello SES developers!

Is it possible to create a compartment with a brand new set of intrinsics?

I understand that lockdown() is meant to remove the need for this but in my use case, the third party code I'm trying to evaluate (google analytics, twitter widgets, etc) frequently performs operations that lockdown() makes impossible. 

Thank you for all the hard work you are putting into this!

**Comment by AidenRourke** [comment]
To illustrate, this is something that the code I'm working with is DETERMINED to do:

```
const list = [1,2,3,4]
print(list.push=someOtherFunction);
```

**Comment by erights** [comment]
> Is it possible to create a compartment with a brand new set of intrinsics?

Hi @AidenRourke that's exactly the distinction between Compartments and Realms, and why both proposals exist and are complementary.

**Comment by erights** [comment]
The problem this example is running into is *not* the need for a new set of intrinsics. It is the override mistake. In your own copy of the SES shim, edit the `%ArrayPrototype%` section of `enablements.js` at
https://github.com/Agoric/SES-shim/blob/master/packages/ses/src/enablements.js#L81
to add the entry
```js
    push: true,
```
and see if that fixes the problem you're running into. Either way, please report the result here and we'll regroup.

**Comment by AidenRourke** [comment]
Hello @erights, thank you for the quick reply.

I am about to try out what you suggested. What does it mean for a property to be repaired?

From what I can tell from the realms-shim repo, work on it has been discontinued. Is this true or has the repo been moved somewhere else?

**Comment by erights** [comment]
> From what I can tell from the realms-shim repo, work on it has been discontinued. Is this true or has the repo been moved somewhere else?

The Agoric realms-shim repo is indeed long stale and should not be used. There is a live Realms proposal at https://github.com/tc39/proposal-realms . Salesforce has been putting by far the most energy into this. 

@caridy @leobalter is there a publicly accessible Realms-shim that others can use? What is the state of this?

**Comment by erights** [comment]
> I am about to try out what you suggested. What does it mean for a property to be repaired?

The [Override Mistake](https://www.google.com/search?sxsrf=ALeKk00edfVoJjMXKdVtEC1MfinYqWK4yw%3A1614663223294&ei=N849YNq9EZKE-gTk-p7QDA&q=%22override+mistake%22+javascript&oq=%22override+mistake%22+javascript&gs_lcp=Cgdnd3Mtd2l6EAMyBAgjECc6BwgjELADECc6BwgAEEcQsANQwkdYrmFgumVoAXACeACAAYABiAGeApIBAzIuMZgBAKABAaoBB2d3cy13aXrIAQnAAQE&sclient=gws-wiz&ved=0ahUKEwiateye8ZDvAhUSgp4KHWS9B8oQ4dUDCA0&uact=5) is demonstrated in raw JS as follows:

```js
function showMistake() {
  "use strict";
  const x = Object.freeze({foo: 8});
  const y = {__proto__: x};
  y.foo = 9;
}
showMistake();

Uncaught TypeError: Cannot assign to read only property 'foo' of object '#<Object>'
    at showMistake (REPL36:5:7)
```

The assignment to `y` should create an own `'foo'` property on `y` that overrides the `'foo'` property that `y` would otherwise inherit from `x`. However, due to one of the worst decisions in the history of the tc39 standards committee, the so-called override mistake, a non-writable *data* property cannot be overridden by assignment. There are two ways out:

If the code with the assignment can be fixed or patched, replace the assignment with a call to `defineProperty`
```js
function showProperFix() {
  "use strict";
  const x = Object.freeze({foo: 8});
  const y = {__proto__: x};
  Object.defineProperty(y, 'foo', { value: 9 });
  return y.foo;
}
showProperFix(); // 9
```
or, to correctly simulate the assignment
```js
function showProperFix() {
  "use strict";
  const x = Object.freeze({foo: 8});
  const y = {__proto__: x};
  Object.defineProperty(y, 'foo', {
    value: 9,
    writable: true,
    enumerable: true,
    configurable: true
  });
  return y.foo;
}
showProperFix(); // 9
```

If you have no control over the code doing the assignment and must run it as is, but you do have control of the definition of the `x.foo` property, then instead of defining it as a data property as shown above, define it instead as an accessor property:
```js
function showEnablement() {
  "use strict";
  const x = Object.freeze({
    get foo() { return 8; },
    set foo(value) {
      if (this === x) { throw Error('cannot mutate x.foo'); }
      Object.defineProperty(this, 'foo', {
        value: 9,
        writable: true,
        enumerable: true,
        configurable: true
      });
    }
  });
  const y = {__proto__: x};
  y.foo = 9;
  return y.foo;
}
showEnablement(); // 9
```
This is the trick that the SES-shim `lockdown` does for those primordial properties listed in the enablements.js file.

**Comment by AidenRourke** [comment]
First of all, it worked! I really appreciate such a prompt and helpful explanation.

Let me make sure I understand this correctly. 

Adding push: true to the enablements allows list objects in the sandbox to reimplement their push method but doesn't allow them to modify to the global array push function.

**Comment by erights** [comment]
> Adding push: true to the enablements allows list objects in the sandbox to reimplement their push method but doesn't allow them to modify to the global array push function.

Exactly correct.

What code is it that adds `push` to array instances in this way? Why? Could you get them to use `defineProperty` instead? Does this arise with just `push` or have you seen the same misbehavior with other properties?

**Comment by AidenRourke** [comment]
Right now I'm testing with the GoogleAnalytics.js script. It's a complex unreadable file that I have slowly, through endowments, given permissions. 

My goal is to be able to fetch the script from their servers and without modifying it and run it in a SES sandbox where it can only interact with my site through restricted channels. Thus, I cannot get them to change their code in any way.

So far, I have only seen this problem with `push`.

**Comment by erights** [comment]
That is great news! Certainly enough justification to add `push` to our enablements. Care to send us a PR including a test case? Ideally demonstrating the arrangement needed for running Google Analytics? Thanks.

**Comment by AidenRourke** [comment]
@erights I'd love to!

**Comment by AidenRourke** [comment]
@erights I'd love to!

**Comment by leobalter** [comment]
> > From what I can tell from the realms-shim repo, work on it has been discontinued. Is this true or has the repo been moved somewhere else?
> 
> The Agoric realms-shim repo is indeed long stale and should not be used. There is a live Realms proposal at https://github.com/tc39/proposal-realms . Salesforce has been putting by far the most energy into this.
> 
> @caridy @leobalter is there a publicly accessible Realms-shim that others can use? What is the state of this?

I have no current Realms-shim to offer. We've been considering changes to the API structure and it's a current WIP.

I believe the closest low level approaches for Realms today would be Node's vm and Browser iframes. Regardless how much the latter's infamy, it's what we have today.

**Comment by erights** [comment]
Closed by #595 
Thanks @AidenRourke !
